### PR TITLE
[KT-23] Add about us feature and functionalities

### PR DESCRIPTION
**Pull request by AudreyFelicio**
A minor frontend bug still persists, but can review the backend stuffs

### CODE REVIEW

File: app/assets/javascripts/about_users.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (donjar) [inline app/assets/javascripts/about_users.js]: Remove this file

File: app/assets/stylesheets/about_users.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AboutUsers controller here.
```
Comment (donjar) [inline app/assets/stylesheets/about_users.scss]: Remove this file

File: app/controllers/about_users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class AboutUsersController < ApplicationController
```
Comment (donjar) [inline app/controllers/about_users_controller.rb:3]: Two spaces

File: app/helpers/about_users_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module AboutUsersHelper
```
Comment (donjar) [inline app/helpers/about_users_helper.rb:3]: Remove this file

File: app/controllers/about_users_controller.rb (hunk 3 above)
Comment (donjar) [reply app/controllers/about_users_controller.rb:3]: Same goes with other newly created files.

File: app/views/home/about.html.haml [hunk 5]
```diff
@@ -14,8 +14,22 @@
     semua orang. Selain itu, kontes ini berisi soal-soal berkualitas
     yang sangat cocok untuk persiapan OSN tingkat kabupaten, provinsi,
     maupun nasional.
+
   %h2 Tim Kami
-  .row#daftar-panitia
-  \... dan banyak lainnya!
-  %h3 Alumni Tim Kami
-  .row#alumni-panitia
+  - team = @about_users.select { |data| !data.is_alumni }
```
Comment (donjar) [inline app/views/home/about.html.haml]: Put more complex logic in either the controller or the helper

File: config/application.rb [hunk 6]
```diff
@@ -32,6 +32,12 @@ class Application < Rails::Application
     config.autoload_paths.push("#{config.root}/lib")
     config.active_job.queue_adapter = :delayed_job
     config.active_record.schema_format = :sql
+    config.before_configuration do
```
Comment (donjar) [inline config/application.rb]: What is this for?

I think we have a dotenv set up in our project already

File: app/views/home/about.html.haml [hunk 7]
```diff
@@ -14,8 +14,22 @@
...
+  - team.each_with_index do |data, index|
+    - base_class = "about-us-person col-md-4 col-sm-8"
+    %div{:class => "#{base_class}", :data => {:name => "#{data.name}", :description => "#{data.description}"}}
```
Comment (donjar) [inline app/views/home/about.html.haml]: You can do something like `%div.#{base_class}` or something I think. Try to look into other HAML files.

File: app/views/users/_show_user_info.html.haml [hunk 8]
```diff
@@ -11,3 +11,33 @@
 - if can? :show_full, @user
   %p #{@user.point} #{point_image}
   /= link_to 'Pasar', market_items_path, class: 'btn'
+
+-if can? :edit, @user.about_user
+  %h2 About Me
+  -if @user.about_user.nil?
+    %p Buat sekarang!
+    = form_for [@user, @user.build_about_user], url: user_about_user_path(@user) do |f|
```
Comment (donjar) [inline app/views/users/_show_user_info.html.haml:18]: Can't you use the `_form` partial here?

**Review by donjar — COMMENTED** [review]
- why are there png files commited in `public/system`?
The files are now untracked from git :)

- don't forget to add tests :)
Will do tonight

- don't forget screenshots as well!
Screenshot 2021-05-13 at 4 46 04 PM
Screenshot 2021-05-13 at 4 46 37 PM

File: config/application.rb (hunk 6 above)
Comment (AudreyFelicio) [reply config/application.rb]: I think this was an issue for my own PC back then and this may only be applicable to me.
Without this I can't run the `rails s` command if I remembered correctly.
I should've git ignore this since this is personal. Thanks for reminding!

Comment (donjar) [reply config/application.rb]: Can you check if removing this now works and if there's alternative solutions? If this is necessary to run stuff we might want to just put this in

File: app/views/users/_show_user_info.html.haml (hunk 8 above)
Comment (AudreyFelicio) [reply app/views/users/_show_user_info.html.haml:18]: Seems like this cannot be done due to `_form` partial which takes in `@user.about_user` instead of `@user.build_about_user`

File: app/views/home/about.html.haml (hunk 7 above)
Comment (AudreyFelicio) [reply app/views/home/about.html.haml]: I think the ones with extra html tags (such as `:data` and `:description`) need to be done in this way, not sure if a better way such as `%div.#{base_class}` exist for those.

File: app/views/users/_show_user_info.html.haml (hunk 8 above)
Comment (donjar) [reply app/views/users/_show_user_info.html.haml:18]: Can look into https://guides.rubyonrails.org/v4.2/layouts_and_rendering.html#using-partials section 3.4.4? Seems like we can pass in variables to the partials.

Can you check if this has been used in some of the other partials we have used? I think we pass local variables to the partials in some way, though I can't really remember how.

File: app/views/home/about.html.haml (hunk 7 above)
Comment (donjar) [reply app/views/home/about.html.haml]: I see. One way is to abstract this thing into a helper. Anyway leaving it like this is also fine, but I think it's better for the class to be put as `.#{base_class}` because it's rather special compared to the data attributes.

File: app/views/users/_show_user_info.html.haml (hunk 8 above)
Comment (AudreyFelicio) [reply app/views/users/_show_user_info.html.haml:18]: Yup, seems like passing in local variables inside partials is the way 🥳 !

File: config/application.rb (hunk 6 above)
Comment (AudreyFelicio) [reply config/application.rb]: Removed this :smile:
The `.env` file works

File: .vagrant/bundler/global.sol [hunk 9]
```diff
@@ -0,0 +1 @@
+{"dependencies":[["log4r",[">= 0"]],["concurrent-ruby",["~> 1.0"]],["i18n",[">= 0"]],["micromachine",[">= 2","< 4"]],["vagrant-vbguest",["= 0.29.0"]]],"checksum":"a4e36ce1ad434d531b6329259c1960a7c0c3f0ef1e2f640bfcd6df4c5cbb544b","vagrant_version":"2.2.14"}
```
Comment (donjar) [inline .vagrant/bundler/global.sol]: Remove this file?

File: .vagrant/rgloader/loader.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+# This file loads the proper rgloader/loader.rb file that comes packaged
```
Comment (donjar) [inline .vagrant/rgloader/loader.rb]: Remove this file?

File: ENGLISH.md [hunk 11]
```diff
@@ -95,11 +95,12 @@ sudo apt install <packagename>
         3. Elevate user privilege: `alter user "ubuntu" with superuser;`
     3. Now we will set the database up with `bundle exec rake db:setup`
     4. Run all migration files by `bundle exec rake db:migrate`
-8. The initial setup to run the server has been done. To run the server in test environment, you can run 
+8. Create `.env` file in the root directory and copy over the contents from `.env.default` to `.env`
```
Comment (donjar) [inline ENGLISH.md]: Should also note that the env vars in `.env` need to be populated with any values that need to be filled in

File: ENGLISH.md [hunk 12]
```diff
@@ -95,11 +95,12 @@ sudo apt install <packagename>
...
+9. The initial setup to run the server has been done. To run the server in test environment, you can run
```
Comment (donjar) [inline ENGLISH.md]: Actually this shouldn't be run in test environment. Should remove the `-e test` as well.

Preferably the readme changes are done in another PR, but putting it here is fine.

File: app/models/about_user.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+class AboutUser < ActiveRecord::Base
```
Comment (donjar) [inline app/models/about_user.rb:19]: This file should have the model annotations in... check the other model files.

The gem is https://github.com/ctran/annotate_models, see if you can make it work?

File: app/views/home/about.html.haml [hunk 14]
```diff
@@ -14,8 +14,47 @@
...
     yang sangat cocok untuk persiapan OSN tingkat kabupaten, provinsi,
     maupun nasional.
+
   %h2 Tim Kami
-  .row#daftar-panitia
-  \... dan banyak lainnya!
-  %h3 Alumni Tim Kami
-  .row#alumni-panitia
+  .row
+    - endFlag = true
```
Comment (donjar) [inline app/views/home/about.html.haml]: Snake case please

File: app/views/home/about.html.haml [hunk 15]
```diff
@@ -14,8 +14,47 @@
...
+    - @team.each_with_index do |data, index|
```
Comment (donjar) [inline app/views/home/about.html.haml]: Can we abstract these complex logic into `app/helpers` folder? You can check the functions there to get an idea of how stuff is organised there.

Ideally we want our views to be rather simple with all the view-related logic to be tucked away there.

File: app/models/user.rb [hunk 16]
```diff
@@ -109,6 +109,9 @@ def before_add_method(role)
 
   has_many :point_transactions
 
+  has_one :about_user, dependent: :destroy
+  accepts_nested_attributes_for :about_user
```
Comment (donjar) [inline app/models/user.rb]: Can you remind me what this is for?

File: app/controllers/home_controller.rb [hunk 17]
```diff
@@ -38,6 +38,9 @@ def about
       image_tag = ActionController::Base.helpers.image_tag(image_path)
       [img_file, image_tag.tr('"', "'")]
     end.to_h
+    @about_users = AboutUser.all.order("RANDOM()")
```
Comment (donjar) [inline app/controllers/home_controller.rb]: If this variable is not passed down to the view, you can just use `about_users`

File: .gitignore [hunk 18]
```diff
@@ -24,10 +24,12 @@
 
 # Sensitive files
 /config/database.yml
+/config/local_env.yml
```
Comment (donjar) [inline .gitignore]: So is this still needed? (and `/vendor/bundle` as well?)

**Review by donjar — COMMENTED** [review]
Generally code looks good! Left some minor comments.

Will test the functionality soon and let you know.

File: app/views/about_users/_form.html.haml [hunk 19]
```diff
@@ -0,0 +1,16 @@
+%p
```
Comment (donjar) [inline app/views/about_users/_form.html.haml]: Can you use the bootstrap classes such as `.form-group` etc? Refer to other forms in the codebase.

File: app/views/about_users/_form.html.haml [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+  = f.label :name, 'Nama:'
+%p
+  = f.text_field :name
```
Comment (donjar) [inline app/views/about_users/_form.html.haml]: Actually, why not just take the name from the user's full name field?

File: app/views/about_users/edit.html.haml [hunk 21]
```diff
@@ -0,0 +1,3 @@
+= form_for [@user, @about_user] do |f|
```
Comment (donjar) [inline app/views/about_users/edit.html.haml]: Should add `.container` as well somewhere in this file. Refer to other pages

**Review by donjar — COMMENTED** [review]
This review is from testing the changes. Should be good to go after all the issues have been resolved.

**Review by donjar — COMMENTED** [review]
- don't forget to delete about_us_entries.js
- please devise some way to migrate the existing about us entries into the database

File: app/models/user.rb (hunk 16 above)
Comment (AudreyFelicio) [reply app/models/user.rb]: The `accepts_nested_attributes_for`?

Comment (AudreyFelicio) [reply app/models/user.rb]: Hmmm I can't seem to remember why did I put this thing...

File: .vagrant/bundler/global.sol (hunk 9 above)
Comment (AudreyFelicio) [reply .vagrant/bundler/global.sol]: Done

File: .vagrant/rgloader/loader.rb (hunk 10 above)
Comment (AudreyFelicio) [reply .vagrant/rgloader/loader.rb]: Done

File: ENGLISH.md (hunk 11 above)
Comment (AudreyFelicio) [reply ENGLISH.md]: Done

File: ENGLISH.md (hunk 12 above)
Comment (AudreyFelicio) [reply ENGLISH.md]: Done

File: app/models/about_user.rb (hunk 13 above)
Comment (AudreyFelicio) [reply app/models/about_user.rb:19]: Yes, I updated the annotations for other models as well

File: app/views/home/about.html.haml (hunk 14 above)
Comment (AudreyFelicio) [reply app/views/home/about.html.haml]: Done

File: app/controllers/home_controller.rb (hunk 17 above)
Comment (AudreyFelicio) [reply app/controllers/home_controller.rb]: 👍

File: .gitignore (hunk 18 above)
Comment (AudreyFelicio) [reply .gitignore]: No, they are deleted now

File: app/views/about_users/_form.html.haml (hunk 20 above)
Comment (AudreyFelicio) [reply app/views/about_users/_form.html.haml]: Take it as the default value or make it permanent?

File: app/views/home/about.html.haml (hunk 15 above)
Comment (AudreyFelicio) [reply app/views/home/about.html.haml]: Yup sure, thanks for letting me know!

File: app/views/about_users/_form.html.haml (hunk 19 above)
Comment (AudreyFelicio) [reply app/views/about_users/_form.html.haml]: 👍

File: app/views/about_users/edit.html.haml (hunk 21 above)
Comment (AudreyFelicio) [reply app/views/about_users/edit.html.haml]: 👍

File: app/models/user.rb (hunk 16 above)
Comment (donjar) [reply app/models/user.rb]: Think it should be safe to remove. I can't find usages of `accepts_nested_attributes_for` in the codebase anyway.

File: app/views/about_users/_form.html.haml (hunk 20 above)
Comment (donjar) [reply app/views/about_users/_form.html.haml]: Hm I think make it permanent. i.e. just don't have the name entry in this model. Just read it from the user model instead.

Comment (AudreyFelicio) [reply app/views/about_users/_form.html.haml]: @donjar So, we assume that the full name is always correct (at least for panitia/admin)?

File: app/models/user.rb (hunk 16 above)
Comment (AudreyFelicio) [reply app/models/user.rb]: 👍

File: app/views/about_users/_form.html.haml (hunk 20 above)
Comment (donjar) [reply app/views/about_users/_form.html.haml]: Yep let's do that.

File: scripts/import.sh [hunk 22]
```diff
@@ -1,7 +1,26 @@
...
+  MINGW*)     machine=MinGw;;
+  *)          machine="UNKNOWN:${unameOut}"
+esac
+if [ "Mac" = "${machine}" ]; then
+  bundle exec rake db:drop db:create
+  createuser ktom
 
-cd import
-cat * | sudo -u postgres psql --set ON_ERROR_STOP=on -1 kontes_terbuka
+  cd import/db_backup/210616_000001
```
Comment (donjar) [inline scripts/import.sh]: This shouldn't be here I think

File: migrate-about-us.rb [hunk 23]
```diff
@@ -1,5 +1,10 @@
-/* exported about_us_data alumni_data */
-var about_us_data = [
+# This is a ruby script to migrate all about us entries
```
Comment (donjar) [inline migrate-about-us.rb]: This script shouldn't be in this folder. Maybe in the `scripts` folder?

File: migrate-about-us.rb [hunk 24]
```diff
@@ -93,21 +98,111 @@
     description: 'Bivan Alzacky Harmanto adalah salah satu siswa yang berhasil mengikuti IMO 2012 dan IMO 2013 dan menyumbangkan medali perunggu untuk Indonesia. Saat ini Bivan sedang menempuh studi S1 jurusan Ilmu Komputer di Korea Advanced Institute of Science and Technology (KAIST), yang merupakan universitas nomor 1 di Korea untuk bidang teknik. Saat ini, Bivan merupakan salah satu anggota tim koreksi KTO Matematika. Selain gemar  mengotak-atik soal-soal Matematika (khususnya geometri) dan komputer, Bivan juga gemar mengecek Facebook tanpa alasan yang jelas (sekadar buka dan lihat-lihat saja), membagi ceritanya di blognya, dan bermain game di laptopnya. Oh iya satu hal lagi, board game sekarang menjadi salah satu kegemarannya, berkat salah satu temannya di KAIST yang terbilang sangat gila terhadap board game.',
     image: 'bivan.jpg',
   },
-];
+]
+
+def get_panitia_or_admin
+  users = User.all
```
Comment (donjar) [inline migrate-about-us.rb]: Can do this on the ActiveRecord layer?

Possibly something like
```ruby
User.joins(:roles).where('role.name' => ['panitia', 'admin'])
```

(the above script might not work, you might need to tweak it)

Doing this on the ActiveRecord layer is advantageous for various reasons. For one, you just execute one SQL query and let the database do what it's supposed to do: filtering through data.

File: migrate-about-us.rb [hunk 25]
```diff
@@ -93,21 +98,111 @@
...
+  end
+end
 
-    var temporary_value = arr[current_index];
-    arr[current_index] = arr[random_index];
-    arr[random_index] = temporary_value;
-  }
-}
+def create_about_users(panitia_or_admin)
+  $about_us_data.map do |about_us|
```
Comment (donjar) [inline migrate-about-us.rb]: Seems like you can abstract these 2 parts together

File: migrate-about-us.rb [hunk 26]
```diff
@@ -93,21 +98,111 @@
...
+  end
+end
 
-randomize_array(about_us_data);
-randomize_array(alumni_data);
+def migrate
+  panitia_or_admin = get_panitia_or_admin()
+  existing_panitia_or_admin, missing_panitia_or_admin = process_about_us(panitia_or_admin)
+  existing_alumni, missing_alumni = process_alumni(panitia_or_admin)
+  create_missing_users(missing_panitia_or_admin)
```
Comment (donjar) [inline migrate-about-us.rb]: Instead of creating missing users with rather arbitrary defaults, let us just print those users that are missing. We can act on them separately (maybe the name is a typo somewhere, etc).

**Review by donjar — COMMENTED** [review]
Be sure to run rubocop & ensure that tests pass as well

File: migrate-about-us.rb (hunk 26 above)
Comment (AudreyFelicio) [reply migrate-about-us.rb]: I see, so we'll create only for those that already exist then? (and perhaps manually ask for those missing to create on their own)

File: scripts/import.sh (hunk 22 above)
Comment (AudreyFelicio) [reply scripts/import.sh]: We need this to specify from where to import right?
Or do we make `/db_backup/210616_000001` as user input instead?

Comment (donjar) [reply scripts/import.sh]: I think the idea is to copy the contents of 210616_000001 directly into the db_backup folder. So no nesting.

File: migrate-about-us.rb (hunk 26 above)
Comment (donjar) [reply migrate-about-us.rb]: Essentially yeah. My concern with us creating the users is that I feel like there are no defaults that can apply to everyone, and if in the end we need to change them anyway, might as well us create them.

File: scripts/migrate-about-us.rb [hunk 27]
```diff
@@ -93,21 +98,75 @@
...
+end
 
-randomize_array(about_us_data);
-randomize_array(alumni_data);
+def migrate
+  panitia_or_admin = get_panitia_or_admin()
+  existing_panitia_or_admin, missing_panitia_or_admin = process_about_us(panitia_or_admin)
+  existing_alumni, missing_alumni = process_alumni(panitia_or_admin)
+  missing_users = missing_panitia_or_admin + missing_alumni
+  all_missing_log = missing_users.reduce() { |acc, curr| acc + "\n" + curr }
```
Comment (donjar) [inline scripts/migrate-about-us.rb]: https://apidock.com/ruby/Array/join use this?

**Review by donjar — APPROVED** [review]
Thanks for your good work!

File: scripts/migrate-about-us.rb [hunk 28]
```diff
@@ -93,21 +98,75 @@
...
+  File.write("./log/missing_users.txt", all_missing_log)
```
Comment (donjar) [inline scripts/migrate-about-us.rb]: I think you can just print this instead of putting it in a file
